docs/src/app/(public)/(content)/react/components/autocomplete/demos/async/css-modules/index.tsx

```diff
@@ -0,0 +1,130 @@
+import * as React from 'react';
+import { Autocomplete } from '@base-ui-components/react/autocomplete';
+import styles from './index.module.css';
+import { top100Movies, type Movie } from './data';
+
+function itemToString(item: Movie) {
+  return item.title;
+}
+
+export default function ExampleAsyncAutocomplete() {
+  const [searchValue, setSearchValue] = React.useState('');
+  const [isLoading, setIsLoading] = React.useState(false);
+  const [searchResults, setSearchResults] = React.useState<Movie[]>([]);
+  const [error, setError] = React.useState<string | null>(null);
+
+  const { contains } = Autocomplete.useFilter({ sensitivity: 'base' });
+
+  const filterByTitleOrYear = React.useCallback(
+    (movie: Movie, query: string) => {
+      return contains(movie.title, query) || contains(movie.year.toString(), query);
+    },
+    [contains],
+  );
+
+  React.useEffect(() => {
+    if (!searchValue) {
+      setSearchResults([]);
+      setIsLoading(false);
+      return undefined;
+    }
+
+    setIsLoading(true);
+    setError(null);
+
+    let ignore = false;
+
+    async function fetchMovies() {
+      try {
+        const results = await searchMovies(searchValue, filterByTitleOrYear);
+        if (!ignore) {
+          setSearchResults(results);
+        }
+      } catch (err) {
+        if (!ignore) {
+          setError('Failed to fetch movies. Please try again.');
+          setSearchResults([]);
+        }
+      } finally {
+        if (!ignore) {
+          setIsLoading(false);
+        }
+      }
+    }
+
+    const timeoutId = setTimeout(fetchMovies, 300);
+
+    return () => {
+      clearTimeout(timeoutId);
+      ignore = true;
+    };
+  }, [searchValue, filterByTitleOrYear]);
+
+  let status: React.ReactNode = `${searchResults.length} result${searchResults.length === 1 ? '' : 's'} found`;
+  if (isLoading) {
+    status = (
+      <React.Fragment>
+        <div className={styles.Spinner} aria-hidden />
+        Searching...
+      </React.Fragment>
+    );
+  } else if (error) {
+    status = error;
+  } else if (searchResults.length === 0 && searchValue) {
+    status = `Movie or year "${searchValue}" does not exist in the Top 100 IMDb movies`;
+  }
+
+  const shouldRenderPopup = searchValue !== '';
+
+  return (
+    <Autocomplete.Root
+      items={searchResults}
+      inputValue={searchValue}
+      onInputValueChange={setSearchValue}
+      itemToString={itemToString}
+      filter={filterByTitleOrYear}
```

Comment:
Sorry for jumping in. Is this prop needed here? I thought the point of async search is to search via API (be it through TanStack Query or `useEffect`), and then the result will be provided via the `items` prop.

Reply:
Are you referring to `filter`? The items are in object format and the internal filter compares strings or `object.value`